File: ctci/chapter_4_trees_graphs/cpp/4.6-successor.cpp
#include "tree.h"
#include "binary_search_tree.h"

/**
 * Traverse to most left child node.
 */
TreeNode *left_most_child(TreeNode *node)
{
    TreeNode *curr = node;
    while (curr->left)
        curr = curr->left;
    return curr;
}

/**
 * Returns the next in-order successive node.
 */
TreeNode *next_successor(TreeNode *node)
{
    /* If node has right child, traverse down and left. */
    if (node->right)
        return left_most_child(node->right);
    TreeNode *child = node;
    TreeNode *parent = child->parent;
    /* Traverse up to first parent whose left child is not original node
     * or previous parent. */
    while (parent && parent->left != child)
    {
        child = parent;
        parent = parent->parent;
    }
    return parent;
}

int main()
{
    BinarySearchTree bst;
    TreeNode *bst5, *bst17;
    TreeNode *s1, *s2;

    bst.insert(20);
    bst.insert(10);
    bst.insert(30);
    bst5 = bst.insert(5);
    bst.insert(3);
    bst.insert(7);
    bst.insert(15);
    bst17 = bst.insert(17);

    s1 = next_successor(bst5);
    cout << s1->value << endl; // 7
    s2 = next_successor(bst17);
    cout << s2->value << endl; // 20

    return 0;
}

File: ctci/chapter_4_trees_graphs/js/4.6-successor.js
const BinarySearchTree = require('./binarySearchTree')
const { TreeNode } = require('./tree')

/**
 * Returns the next in-order successive node.
 * @param   {TreeNode}  node
 * @returns {TreeNode}
 */
const nextSuccessor = node => {
  /* If node has right child, traverse down and left. */
  if (node.right) return leftMostChild(node.right)
  let child = node
  let parent = child.parent
  /* Traverse up to first parent whose left child is not original node or
   * previous parent. */
  while (parent && parent.left != child) {
    child = parent
    parent = parent.parent
  }
  return parent
}

/**
 * Traverse to most left child node.
 * @param   {TreeNode}  node
 * @returns {TreeNode}
 */
const leftMostChild = node => {
  while (node.left) node = node.left
  return node
}

bst = new BinarySearchTree()
bst.insert(20)
bst.insert(10)
bst.insert(30)
let bst5 = bst.insert(5)
bst.insert(3)
bst.insert(7)
bst.insert(15)
let bst17 = bst.insert(17)

console.log(nextSuccessor(bst5))
/*
TreeNode {
  value: 7,
  left: null,
  right: null,
  parent: ... }
*/
console.log(nextSuccessor(bst17))
/*
TreeNode {
  value: 20,
  left:
   TreeNode {
     value: 10,
     left:
      TreeNode {
        value: 5,
        left: [TreeNode],
        right: [TreeNode],
        parent: [Circular] },
     right:
      TreeNode { value: 15, left: null, right: [TreeNode], parent: [Circular] },
     parent: [Circular] },
  right:
   TreeNode { value: 30, left: null, right: null, parent: [Circular] },
  parent: null }
 */
